rl_bakery/agent_trainer/agents.py

```diff
@@ -0,0 +1,173 @@
+import abc
+import tensorflow as tf
+from tf_agents.utils import common
+from tf_agents.networks import q_network
+from tf_agents.agents.dqn import dqn_agent
+from tf_agents.agents import ddpg
+from tf_agents.trajectories import time_step as ts
+import logging
+
+from rl_bakery.agent_trainer.config import Optimizer
+
+logger = logging.getLogger(__name__)
+
+
+class AgentTrainer(metaclass=abc.ABCMeta):
+    """
+    This set ups an Agent to be trained.
+    """
+
+    def __init__(self, data_spec, config):
+        self._data_spec = data_spec
+        self._config = config
+
+    @abc.abstractmethod
+    def init_agent(self):
+        pass
+
+
+class TFAgentTrainer(AgentTrainer):
+
+    def __init__(self, data_spec, config):
+        super().__init__(data_spec, config)
+
+    @abc.abstractmethod
+    def _init_tfagent(self, optimizer):
+        pass
+
+    def _get_optimizer(self, optimizer_config):
+
+        # TODO use tf2 optimizers, support other optimizers
+        if optimizer_config.optimizer == Optimizer.Adam:
+            return tf.compat.v1.train.AdamOptimizer(learning_rate=optimizer_config.learning_rate)
+        elif optimizer_config.optimizer == Optimizer.RMSprop:
+            return tf.compat.v1.train.RMSPropOptimizer(learning_rate=optimizer_config.learning_rate)
+        else:
+            raise Exception("Unknown optimizer %s" % optimizer_config.optimizer)
+
+    def init_agent(self):
+        # get the global step
+        global_step = tf.compat.v1.train.get_or_create_global_step()
+        return self._init_tfagent(global_step)
+
+
+class QAgent(TFAgentTrainer):
+
+    def __init__(self, data_spec, config):
+        super().__init__(data_spec, config)
+
+    def _build_q_network(self):
+        # By default, this builds a q-network using the supplied fc_layer_params
+        # This function can be overridden to provide a more complicated Q Network
+        assert(self._config.agent.fc_layer_params)
+        q_net = q_network.QNetwork(self._data_spec.observation_spec, self._data_spec.action_spec,
```

Comment:
Same here in relation to the params of QNetwork such as dropout, etc.

https://github.com/tensorflow/agents/blob/master/tf_agents/networks/q_network.py#L46

Reply:
it's okay, for more complex options they can override the function themselves